pkg/operator/ceph/cluster/osd/osd.go

```diff
@@ -193,16 +195,79 @@ func (c *Cluster) Start() error {
 	}
 	logger.Infof("wait timeout for healthy OSDs during upgrade or restart is %q", c.clusterInfo.OsdUpgradeTimeout)
 
+	osdsToSkipReconcile, err := c.getOSDsToSkipReconcile()
+	if err != nil {
+		logger.Warningf("failed to get osds to skip reconcile. %v", err)
+	}
+
 	// prepare for updating existing OSDs
 	updateQueue, deployments, err := c.getOSDUpdateInfo(errs)
 	if err != nil {
 		return errors.Wrapf(err, "failed to get information about currently-running OSD Deployments in namespace %q", namespace)
 	}
-	osdsToSkipReconcile, err := c.getOSDsToSkipReconcile()
-	if err != nil {
-		logger.Warningf("failed to get osds to skip reconcile. %v", err)
+
+	var replaceConfig *OSDReplaceConfig
+	if c.spec.Storage.Store.UpdateStore == OSDStoreUpdateConfirmation {
+		logger.Infof("start migrating osds to backend store %q", c.spec.Storage.Store.Type)
+
+		replaceConfig, err = c.getOSDReplaceConfig()
+		if err != nil {
+			return errors.Wrapf(err, "failed to get OSD replace config")
+		}
+
+		// skip updating the OSDs that need migration. This is to prevent the reconciler from updating the backend store in the OSD deployments
+		// that have not been migrated to the new store yet.
+		for _, osd := range replaceConfig.OSDs {
+			updateQueue.Remove([]int{osd.ID})
+		}
+
+		if len(replaceConfig.OSDs) != 0 {
+			logger.Infof("total osds that need migration to new backend store %q : %d", c.spec.Storage.Store.Type, len(replaceConfig.OSDs))
+			pgHealthMsg, pgClean, err := cephclient.IsClusterClean(c.context, c.clusterInfo)
```

Comment:
Instead of returning when the PGs are not clean and immediately failing the reconcile, I'm thinking we should wait for some time period for the PGs to come clean. For example, check every 10s, and only if the PGs are still not clean after 15min then we can fail the reconcile and try again. Then we can reduce the number of reconciles where we start completely over. For OSDs with not much backfilling, this also makes the conversion more likely to succeed for each OSD without failing the reconcile.

Reply:
Oh now I see that if they are not clean, we just allow the reconcile to continue for other OSDs. after the warning on line 263. Instead of my suggestion to wait for clean PGs before starting the next OSD replacement, maybe it's better to wait for clean PGs immediately after completing an OSD replacement, so it won't block other OSD reconcile.

Reply:
are suggesting this flow:
- check if pgs are clean
- Delete osd if pgs are clean, else skip osd replacement
- complete the reconcile of other osds
- check again if the pgs are clean after all osds have reconciled. If not clean, then keep retrying every 10 seconds for next 15 minutes. Continue with the reconcile if pgs become healthy within this time. Else, return error and reconcile?

Reply:
Yes that flow looks good, thanks